### PR TITLE
Create E2E testing for receiver logic.

### PR DESCRIPTION
**Pull request by hbulgarini**
Related to #191

This PR extends current E2E testing with the following enhancements

There are now 4 testing asserts:

Test 1: Should provide native source address type and should show native address in receiver field.
Test 2: Should provide companion source address type and should show companion address in receiver field.
Test 2: Should provide generic address type and should show two options,  choose native and get native receiver field.
Test 4: Should execute end to end testing of transaction. This includes the following checks:

- Waits for transaction broadcasted message.
- Check that all steps get completed.
- Checks that balance is updated accordingly.

### CODE REVIEW

File: e2e/app.test.js [hunk 1]
```diff
@@ -0,0 +1,274 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
```
Comment (wirednkod) [inline e2e/app.test.js:1]: I do not understand why this test exist? It launches a browser and runs tests just like the puppeteer E2E:alone test does.
IMO we should have 3 tests.
1) functionality tests (you already have build those)
2) E2E test (1 happy path test that shows that flow works - already built those)
3) component tests (these does not have to open a browser in order to run)

The way this is implemented IMO is not correct as it takes way too much time to run (open browser, wait, run, etc etc) and it should be able to run also in CI (which this kind of tests are not able i think)

File: package.json [hunk 2]
```diff
@@ -50,8 +50,8 @@
     "prestart": "yarn downloadCustomTypes",
     "start": "PORT=3001 react-app-rewired start",
     "build": "react-app-rewired build",
-    "test:e2e": "jest -c e2e/jest.config.js",
-    "test:e2e-w": "jest -c e2e/jest.config.js --watch",
+    "test:e2e": "jest -c e2e/jest.config.js --detectOpenHandles",
```
Comment (wirednkod) [inline package.json:53]: I think `--detectOpenHandles` should be used only when we want to identify leaks (and Jest informs about them anyway) not in each run

File: e2e/app.test.js (hunk 1 above)
Comment (hbulgarini) [reply e2e/app.test.js:1]: Well i think that it does not have to be so drastic... the additional tests added do not add more than 30 seconds or similar so even if it adds minutes to the process is something we can deal with more taking into account that the gain is significant in terms of consistency and quality.

Then I agree with your three suggestions but the purpose of this PR is to extend and cover all the critical path of the transfer feature and the complex logic around the receiver field that needs to work perfectly and this tests validate that.

The reason for including the tests on this jest file is because we can combine puppeter tools with powerful jest assertions which i think it's the best of two worlds.

Finally , the next step is to include these tests to the CI so we can run these tests in every deploy of master branch: #192

Comment (wirednkod) [reply e2e/app.test.js:1]: IMO all test should not take more than 20 seconds. In Substrate connect, covering 2 major projects with approx 80-90% coverage of unit tests takes 20 seconds. 
In addition i do not understand how exactly these tests will run in CI

Comment (hbulgarini) [reply e2e/app.test.js:1]: Ok fair enough ... 

The goal for the tests are described in the description of this PR:

Test 1: Should provide native source address type and should show native address in receiver field.
Test 2: Should provide companion source address type and should show companion address in receiver field.
Test 2: Should provide generic address type and should show two options, choose native and get native receiver field.
Test 4: Should execute end to end testing of transaction. This includes the following checks:

Comment (wirednkod) [reply e2e/app.test.js:1]: I know. I guess what i am trying to say is that these tests can run without even initiating a browser as far as I know

File: e2e/jest.config.js [hunk 3]
```diff
@@ -18,9 +18,9 @@ module.exports = {
   preset: 'jest-puppeteer',
   globals: {
     SERVER_URL: 'http://localhost:3001',
-    JEST_TIMEOUT: 10000
+    JEST_TIMEOUT: 100000
```
Comment (tomusdrw) [inline e2e/jest.config.js:21]: Is it because the 4th test takes so long?

File: package.json (hunk 2 above)
Comment (tomusdrw) [reply package.json:53]: What's the drawback of using it always?

File: src/components/Icons.tsx [hunk 4]
```diff
@@ -44,15 +44,15 @@ interface IconApiStatusProps {
 export const IconTxStatus = ({ status }: IconTxStatusProps) => {
   switch (status) {
     case TransactionStatusEnum.COMPLETED:
-      return <CheckCircleOutlineIcon />;
+      return <CheckCircleOutlineIcon id="check-circle-icon" />;
```
Comment (tomusdrw) [inline src/components/Icons.tsx:47]:
```suggestion
      return <CheckCircleOutlineIcon id="test-transaction-status-completed" />;
```
What's the point of bounding this to a concrete component type? 

I'd also distinguish for-tests ids by using `test` prefix everywhere.

File: e2e/app.test.js (hunk 1 above)
Comment (tomusdrw) [reply e2e/app.test.js:1]: We shouldn't be adding any more E2E tests for now, but the ones that are here can stay. The focus should be more on unit testing, until we get some major features (like a new tab). 

Running it on the CI is a priority (at least after merging to master and deploying).

> I guess what i am trying to say is that these tests can run without even initiating a browser as far as I know

What's the alternative? Does pupetteer have some built-in browser?

File: e2e/app.test.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+};
+
+const timeout = 500000;
+
+const ids = {
+  native: '#test-native-input',
+  companion: '#test-companion-input',
+  sender: '#test_sender_component',
+  transferButton: '#test-button-submit',
+  senderComponent: '#test_sender_component'
```
Comment (tomusdrw) [inline e2e/app.test.js:54]: Consistency please!!! Either `_` or `-` everywhere.

File: package.json (hunk 2 above)
Comment (wirednkod) [reply package.json:53]: 2 drawbacks:
a) It implies `--runInBand`, making tests run serially.
b) It has a significant performance penalty

Documentation proposes that this should only be used for debugging.

File: e2e/app.test.js (hunk 1 above)
Comment (wirednkod) [reply e2e/app.test.js:1]: > We shouldn't be adding any more E2E tests for now, but the ones that are here can stay. 

I agree. As soon as the work is done - there is no need for removing or anything.

> Running it on the CI is a priority

Im not sure how one can run GUI tests in Github CI - I do not say that it cannot happen - I just do not know how (never done it)

> Does puppetteer have some built-in browser

Puppeteer's browsers can be downloaded along with it (meaning you can add chromium and/or firefox with the installation of the package and run the tests with the downloaded browsers (cross-browsing testing). 
The configuration I did, does not include any browser and it is assumed that chrome is installed with it.

On top of that, puppeteer can run headless browsers (without the need of any GUI) and that is my point above. Lets run the E2E or component tests, but not with full browsers and without such a waiting time. 

I would suggest - since new components implementation start - to pause unit (component tests for now) and implement them whenever a new component is created (e.g. build a new dropdown -> write also unit/component tests for it on the way).

Comment (hbulgarini) [reply e2e/app.test.js:1]: @wirednkod the browser setup is something i plan to work on #192, this PR only includes the new tests described before.

File: e2e/app.test.js (hunk 5 above)
Comment (hbulgarini) [reply e2e/app.test.js:54]: omg!

File: package.json (hunk 2 above)
Comment (hbulgarini) [reply package.json]: Right. It is not needed .. i have removed the parameter.

File: e2e/jest.config.js (hunk 3 above)
Comment (hbulgarini) [reply e2e/jest.config.js]: not really, it can remain as it was before. I rolled back the change.

File: src/components/Icons.tsx (hunk 4 above)
Comment (hbulgarini) [reply src/components/Icons.tsx]: I use this component to validate in the UI if the step was completed or not.

Comment (tomusdrw) [reply src/components/Icons.tsx]: > I use this component to validate in the UI if the step was completed or not.

I understand that. My point is that you are relying on a very specific component to be displayed in something that is a high-level test. The point is that if we decide to change the icon you will need to update the `id` and all references because the id is so tightly coupled with the component being displayed.
Also imagine that we want to write a new test (or assertion). To understand what the test is doing, you need to understand what is `test-check-circle-icon` component and in what circumstances it's display (pretty much the only way to do it so to go to the component and read the code), after doing that you will actually know that: "Aha, that component is displayed when the transaction is completed, so let me use the id.".

So long story short, by using `ids` like that:
1. You are tightly coupling specific design choice (circle icon component) with high-level application logic (transaction is completed). Changing designs will cause extra changes in completely unrelated parts of code.
2. You are creating additional mental overhead to understand what "circle icon component" actually means (transaction is completed).

**Review by tomusdrw — APPROVED** [review]

File: src/components/Icons.tsx (hunk 4 above)
Comment (hbulgarini) [reply src/components/Icons.tsx]: Ok, fair enough ... i will switch it to `test-transaction-step-completed` then
